### PR TITLE
Allow for one-off query overrides for serverside rendering

### PR DESCRIPTION
**Pull request by toddtarsi**
First off, thank you guys for making this. This codebase is really awesome, and I love the approach to everything. Currently though, the main known open source serverside rendering solution (isomorphic-relay, another great codebase) is kind of in a bad spot when it comes to credential passing.

An example of this can be seen with isomorphic-relay, where the solution for passing httponly cookies in serverside rendering involves converting the endpoint to run sequentially, which is a major deoptimizer. By exposing overrides like this, we could skirt that approach by instead just passing in one off data into the sendQueries call.

Admittedly, I don't know this codebase all too well, and there are probably a lot of gotchas I'm overlooking, but I wanted to put this out there and just see what your guys thoughts were. Also, if you just want to see more polish and consistency, let me know and I can refactor / rebase this as needed.

Thanks as always!

### CODE REVIEW

**Comment by facebook-github-bot-7** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by josephsavona** [comment]
Thanks for the PR. We're open to considering this, but can you provide more detail as to why this is needed and how it would be used? It would be helpful to have a link to an example of how `isomorphic-relay` is forced to do serial fetching w/o this.

**Comment by facebook-github-bot-2** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by toddtarsi** [comment]
Absolutely, thank you for the prompt response:

This issue is closed in isomorphic-relay-router, but a quick read will help explain the issue:

https://github.com/denvned/isomorphic-relay-router/issues/13

By exposing this code in relay, and then adding an extra argument for overrides to the prepareData operation in isomorphic-relay-router and isomorphic-relay, we could pass through cookie data to the sendQueries call that these wrap without the overhead of the solution above.

**Comment by toddtarsi** [comment]
Ha, on second thought, that issue tracker rambles:
TLDR: Since node-fetch proxies requests, and there isn't a one off way to override these, serverside rendering hits a challenge where it needs to override this using injectNetworkLayer. Since injectNetworkLayer is a global, the only current solution is to guard from async attack vectors by converting the stack to run sequentially.
By passing in an override function, it gives us a method to customize each request batch without injecting a new global network layer each time on the server.

**Comment by toddtarsi** [comment]
Hey all, after talking with @robrichard, it sounds like my use case was resolved when Relay.environment was able to be used non-globally.
https://github.com/denvned/isomorphic-relay/pull/43
Thanks for taking the time to look this over, and if you'd like I can close this pull request now.
My use cases are resolved I believe.

**Comment by toddtarsi** [comment]
Absolutely, thanks for the quick responses from everyone. 😄
